### PR TITLE
Add innerException to ResourceReader FileNotFoundException

### LINKED ISSUE #48159: net 5 Cannot read resources that depend on serialization
Hi,
We moved to net 5.0.3 (from net 5) and now we are getting thea above exception in Winform form in the Designer on lines like:

this.Icon = ((System.Drawing.Icon)(resources.GetObject("$this.Icon")));
this.btnOpenProtocolsFolder.Image = ((System.Drawing.Image)(resources.GetObject("btnOpenProtocolsFolder.Image")));

what could e the reason?

### PR DESCRIPTION
**Pull request by joperezr**
Fixes https://github.com/dotnet/runtime/issues/48159

cc: @ericstj @filipnavara

### CODE REVIEW

**Review by filipnavara — APPROVED** [review]
Thanks!

**Review by danmoseley — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs [hunk 1]
```diff
@@ -981,9 +981,9 @@ private Type UseReflectionToGetType(int typeIndex)
             //
             // We don't want to regress the expected case by checking the type info before
             // getting to Type.GetType -- this is costly with v1 resource formats.
-            catch (FileNotFoundException)
+            catch (FileNotFoundException fileNotFoundException)
             {
-                throw new NotSupportedException(SR.NotSupported_ResourceObjectSerialization);
+                throw new NotSupportedException(SR.NotSupported_ResourceObjectSerialization, fileNotFoundException);
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs:986]:
```suggestion
                // Include the FileNotFoundException as an inner exception to make it more diagnosable
                throw new NotSupportedException(SR.NotSupported_ResourceObjectSerialization, fileNotFoundException);
```
?

File: src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs [hunk 2]
```diff
@@ -981,9 +981,10 @@ private Type UseReflectionToGetType(int typeIndex)
             //
             // We don't want to regress the expected case by checking the type info before
             // getting to Type.GetType -- this is costly with v1 resource formats.
-            catch (FileNotFoundException)
+            catch (FileNotFoundException fileNotFoundException)
             {
-                throw new NotSupportedException(SR.NotSupported_ResourceObjectSerialization);
+                // Include the FileNotFoundException as an inner exception to make it more diagnosable
+                throw new NotSupportedException(SR.NotSupported_ResourceObjectSerialization, fileNotFoundException);
```
Comment (ericstj) [inline src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs:987]: Should we also change the message here to indicate something other than not-supported?  For reference here's the source of this exception in .NETFramework.

Looks like this code was previously only throwing `NotSupportedException` on platforms where FEATURE_SERIALIZATION was not set, but now we support serialization.  I wonder if we should bring back the ifdef, but instead as a runtime check of [_permitDeserialization](https://source.dot.net/System.Private.CoreLib/R/40812d5433974bf9.html)

In other words
```c#
if (_permitDeserialization) 
    throw;

throw new NotSupportedException(SR.NotSupported_ResourceObjectSerialization, fileNotFoundException);
```

**Comment by ericstj** [comment]
> Add innerException to ResourceReader when it doesn't support serialization

This is a misnomer.  We actually do support deserialization in this case, but the code from when we didn't was left behind.  Have a look at my suggestion.

File: src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs (hunk 2 above)
Comment (ericstj) [reply src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs:987]: Also, it was likely me who missed realizing we needed to undo that ifdef here: https://github.com/dotnet/coreclr/commit/791332b101524034ca9a69ab7669d5a87713447b

I wonder if there are any other interesting `FEATURE_SERIALIZATION` ifdefs that are causing issues.

Comment (joperezr) [reply src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs:987]: I suppose you mean this? https://github.com/dotnet/coreclr/commit/791332b101524034ca9a69ab7669d5a87713447b#diff-6722a5dffc850c4a6d80bc6880484092e7f159fe79746d04ed74440a16d9f348R764-R767

Comment (joperezr) [reply src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs:987]: Actually, this won't work for System.Resources.Extensions library. `_permitDeserialization` is only defined in corelib's resource reader, but not on the extensions one. I could ifdef the condition to only be applicable in corelib

**Review by stephentoub — APPROVED** [review]

**Review by ericstj — APPROVED** [review]
LGTM, thanks for the fix!
